状態変更系

# Issue: パッシブ 能力対象拡張

## 応答のルール
- 常に日本語で応答してください。  
- コードや設定ファイルはそのままのフォーマットで提示してください。  

## プロジェクト概要
- **実行環境**：AWS Lambda（AppSync GraphQL リゾルバ）  
- **言語**：Python 3.x  
- **主要機能**：  
  - DynamoDB によるマッチ状態管理  
  - 各種バトルアクションのディスパッチ  
  - リーダー「パッシブオーラ」の付与／解除  

## ディレクトリ規約
| フォルダ・ファイル          | 用途                                                    |
| --------------------------- | ------------------------------------------------------- |
| `lambda_function.py`        | エントリポイント（AppSync → DynamoDB → イベント解決） |
| `actions/`                  | 各種アクションモジュール（draw, move_zone, aura など）  |
| `helper.py`                 | 共通ユーティリティ（ステータス操作、ターゲット解決等） |
| `schema.graphql`            | GraphQL スキーマ定義                                    |

## 基本作業方針
1. **Issue 作成・レビュー**  
   - 仕様を明確化し、不明点を質問  
2. **Plan（設計）**  
   - `Assets/DocsForAI/Plan/` 相当のフォルダに Markdown で落とし込む  
3. **Implement**  
   - `lambda_function.py` / `actions/` 配下に実装  
   - 単一責任の原則を守る  
4. **Design ドキュメント**  
   - 実装後、`Assets/DocsForAI/Design/` に設計書としてまとめる  
5. **PR & マージ**  
   - PR を作成、レビュー → ユーザー（オーナー）がマージ  

---

# Issue: パッシブアビリティの適用対象拡張

## 目的
- `refresh_passive_auras` の適用範囲を「フィールド上のカードのみ」から「指定されたすべてのゾーン（例：Field、Environment、Counterなど）」に拡張し、あらゆるカードにリーダーパッシブを付与／解除できるようにする。

## 要件
1. **対象ゾーン拡張**  
   - 各アクション定義の `target`（例：`PlayerField`、`Environment`）に応じて、該当ゾーン上のカードすべてを対象にする。  
2. **常時再評価**  
   - `refresh_passive_auras` が呼ばれるたびに最新のゲーム状態をもとに付与／解除を行う。  
3. **全プレイヤー・全ステージ対応**  
   - 全プレイヤーのリーダー定義から進化ステージごとの `passiveEffects` を取得し、各ステージのパッシブを適用／解除する。

## 技術的制約
- **単一責任の原則**を厳守し、  
  - ゾーンフィルタリング  
  - 条件評価（`evaluate_condition`）  
  - 付与処理（`apply_passive_effect`）  
  - 解除処理（`clear_passive_from_targets`）  
  をそれぞれ別メソッドに分割する。  
- パフォーマンスを考慮し、**不必要な全カード走査**は行わない。

## 開発ガイドライン
- **定数化**：対象ゾーンリストを定数として切り出す。  
- **ユーティリティ活用**：既存の `resolve_targets(dummy, act, item)` を使って、アクション定義の `target` にマッチするカード一覧を取得。  
- **メソッド構成**：  
  1. `refresh_passive_auras(item, events)`  
  2. `apply_passive_effect(eff, player, item, events)`  
  3. `clear_passive_from_targets(eff, player, item, events)`  
- 新しい条件式追加時にも対応しやすいよう、`evaluate_condition` にフックを用意する。

## タスク
- [ ] 対象ゾーンリスト（例：`TARGET_ZONES = ["Field","Environment","Counter",…]`）を定義  
- [ ] `refresh_passive_auras` にゾーンフィルタリングを追加  
- [ ] `apply_passive_effect` / `clear_passive_from_targets` のリファクタ  
- [ ] READMEに記載した `matchItem` を使って動作イメージをドキュメント化  
- [ ] 	Phase2 の PowerAura target=PlayerField は A のみ付与
- [ ] Phase3 の DamageAura target=Environment があれば B に付与

## 補足
今後、EnemyFieldCount>=2 のような複雑条件追加に備え、evaluate_condition の拡張ポイントを明示しておく。

@claude
追加タスク
1
下記のエラーが出てた。
ターンが帰ってきた時に。
attackerとかはデータの中にない場合もあるから、安全に処理して
[ERROR] Exception: invalid attacker
Traceback (most recent call last):
  File "/var/task/lambda_function.py", line 709, in lambda_handler
    raise Exception("invalid attacker")

2
apply_passive_effectで付与した能力が、条件を満たさなくなったら解除するっていう要件も追加で入れて
解除するのは再評価（refresh_passive_auras）のタイミングでいい

--- a/lambda_function.py
+++ b/lambda_function.py
@@ -572,11 +572,33 @@ def lambda_handler(event, context):
     if field=="publishClientUpdate":
         return {**args, "timestamp": now_iso()}
 
-    # マッチ読み込み
-    mid=args.get("matchId") or args.get("id")
-    if not mid: raise Exception("need matchId/id")
-    item=table.get_item(Key={"pk":mid,"sk":"STATE"}).get("Item")
-    if not item: raise Exception("match not found")
+    # マッチ読み込み - 安全な処理
+    mid = args.get("matchId") or args.get("id")
+    if not mid:
+        # マッチIDがない場合は適切なエラーレスポンスを返す
+        return {
+            "match": None,
+            "events": [{
+                "type": "MissingMatchId",
+                "payload": {
+                    "message": "マッチIDが必要です"
+                }
+            }]
+        }
+    
+    item = table.get_item(Key={"pk": mid, "sk": "STATE"}).get("Item")
+    if not item:
+        # マッチが見つからない場合は適切なエラーレスポンスを返す
+        return {
+            "match": None,
+            "events": [{
+                "type": "MatchNotFound",
+                "payload": {
+                    "matchId": mid,
+                    "message": "指定されたマッチが見つかりません"
+                }
+            }]
+        }
 
     # -------- getMatch ----------------------------------------
     if field == "getMatch":
@@ -612,10 +634,33 @@ def lambda_handler(event, context):
     # -------- summonCard --------------------------------------
     if field=="summonCard":
         print(f"SummonCard: {args}")
-        cid=args["cardId"]
-        card=next((c for c in item["cards"] if c["id"]==cid),None)
-        if not card: raise Exception("card not found")
-        if card["zone"]=="Field": raise Exception("already on field")
+        cid = args.get("cardId")
+        if not cid:
+            # カードIDがない場合は何もしない
+            return {"match": json.loads(json.dumps(item, cls=DecimalEncoder)), "events": []}
+        
+        card = next((c for c in item["cards"] if c["id"]==cid), None)
+        if not card:
+            # カードが見つからない場合はエラーイベントを返す
+            error_event = {
+                "type": "CardNotFound",
+                "payload": {
+                    "cardId": cid,
+                    "message": "指定されたカードが見つかりません"
+                }
+            }
+            return {"match": json.loads(json.dumps(item, cls=DecimalEncoder)), "events": [error_event]}
+        
+        if card["zone"] == "Field":
+            # 既にフィールドにある場合はエラーイベントを返す
+            error_event = {
+                "type": "AlreadyOnField",
+                "payload": {
+                    "cardId": cid,
+                    "message": "カードは既にフィールドに存在します"
+                }
+            }
+            return {"match": json.loads(json.dumps(item, cls=DecimalEncoder)), "events": [error_event]}
         detach_auras(card,item["cards"])
         card["zone"]="Field"
         trig=[{"type":"OnSummon","payload":{"cardId":cid}},
@@ -699,19 +744,39 @@ def lambda_handler(event, context):
     
     # --- declareAttack -----------------------------------
     if field == "declareAttack":
-        cid_a = args["attackerId"]
+        cid_a = args.get("attackerId")
         cid_t = args.get("targetId")
-        is_leader = args["targetIsLeader"]
+        is_leader = args.get("targetIsLeader", False)
 
-        # 1) フィールドチェック
+        # 1) フィールドチェック - 安全な処理
+        if not cid_a:
+            # 攻撃者IDがない場合は何もしない
+            return {"match": json.loads(json.dumps(item, cls=DecimalEncoder)), "events": []}
+        
         attacker = find_card(item, cid_a)
         if not attacker or attacker["zone"] != "Field":
-            raise Exception("invalid attacker")
+            # 攻撃者が無効な場合はエラーイベントを返す
+            error_event = {
+                "type": "InvalidAttacker",
+                "payload": {
+                    "attackerId": cid_a,
+                    "message": "攻撃者が無効です（存在しないかフィールドにいません）"
+                }
+            }
+            return {"match": json.loads(json.dumps(item, cls=DecimalEncoder)), "events": [error_event]}
 
         if not is_leader:
             target = find_card(item, cid_t)
             if not target or target["zone"] != "Field":
-                raise Exception("invalid target")
+                # ターゲットが無効な場合はエラーイベントを返す
+                error_event = {
+                    "type": "InvalidTarget",
+                    "payload": {
+                        "targetId": cid_t,
+                        "message": "ターゲットが無効です（存在しないかフィールドにいません）"
+                    }
+                }
+                return {"match": json.loads(json.dumps(item, cls=DecimalEncoder)), "events": [error_event]}
         else:
             target = None
 
@@ -783,13 +848,30 @@ def lambda_handler(event, context):
     if field == "setBlocker":
         bid = args.get("blockerId")      # None = ノーブロック
         pb  = item.get("pendingBattle") or {}
+        
+        # バトルステップの安全な確認
         if item.get("battleStep") != "BlockChoice":
-            raise Exception("Not in BlockChoice")
+            error_event = {
+                "type": "InvalidBattleStep",
+                "payload": {
+                    "currentStep": item.get("battleStep"),
+                    "message": "ブロック選択段階ではありません"
+                }
+            }
+            return {"match": json.loads(json.dumps(item, cls=DecimalEncoder)), "events": [error_event]}
 
         if bid:
             blk = find_card(item, bid)
-            if not blk or blk["ownerId"] == pb["attackerOwnerId"]:
-                raise Exception("invalid blocker")
+            if not blk or blk["ownerId"] == pb.get("attackerOwnerId"):
+                # 無効なブロッカーの場合はエラーイベントを返す
+                error_event = {
+                    "type": "InvalidBlocker",
+                    "payload": {
+                        "blockerId": bid,
+                        "message": "無効なブロッカーです（存在しないか攻撃者と同じプレイヤーです）"
+                    }
+                }
+                return {"match": json.loads(json.dumps(item, cls=DecimalEncoder)), "events": [error_event]}
         pb["blockerId"] = bid
         item["pendingBattle"] = pb
         item["battleStep"]    = "AttackAbility"
@@ -806,7 +888,14 @@ def lambda_handler(event, context):
     if field == "resolveBattle":
         # AttackAbility → Resolve をクライアントが要求
         if item.get("battleStep") != "AttackAbility":
-            raise Exception("Not in AttackAbility")
+            error_event = {
+                "type": "InvalidBattleStep",
+                "payload": {
+                    "currentStep": item.get("battleStep"),
+                    "message": "アビリティ発動段階ではありません"
+                }
+            }
+            return {"match": json.loads(json.dumps(item, cls=DecimalEncoder)), "events": [error_event]}
 
         events = []
         resolve_battle(item, events)      # Destroy / Damage を積む
@@ -879,5 +968,14 @@ def lambda_handler(event, context):
                 "turnPlayerId": item.get("turnPlayerId"),
                 "updatedAt": item["updatedAt"]}
 
-    # 未サポート
-    raise Exception(f"Unsupported field: {field}")
+    # 未サポート - 安全な処理
+    return {
+        "match": json.loads(json.dumps(item, cls=DecimalEncoder)),
+        "events": [{
+            "type": "UnsupportedField",
+            "payload": {
+                "field": field,
+                "message": f"サポートされていないフィールドです: {field}"
+            }
+        }]
+    }

--- a/test_safe_error_handling.py
+++ b/test_safe_error_handling.py
@@ -1,0 +1,135 @@
+#!/usr/bin/env python3
+"""
+安全なエラーハンドリングのテストファイル
+"""
+import json
+import boto3
+from moto import mock_dynamodb
+from decimal import Decimal
+from lambda_function import lambda_handler
+
+@mock_dynamodb
+def test_safe_error_handling():
+    """安全なエラーハンドリングのテスト"""
+    
+    # DynamoDB テーブルのセットアップ
+    dynamodb = boto3.resource('dynamodb', region_name='us-east-1')
+    table = dynamodb.create_table(
+        TableName='BattleSync',
+        KeySchema=[
+            {'AttributeName': 'pk', 'KeyType': 'HASH'},
+            {'AttributeName': 'sk', 'KeyType': 'RANGE'}
+        ],
+        AttributeDefinitions=[
+            {'AttributeName': 'pk', 'AttributeType': 'S'},
+            {'AttributeName': 'sk', 'AttributeType': 'S'}
+        ],
+        BillingMode='PAY_PER_REQUEST'
+    )
+    
+    # テスト用のマッチデータ
+    test_match = {
+        'pk': 'test-match-001',
+        'sk': 'STATE',
+        'cards': [
+            {
+                'id': 'card-001',
+                'zone': 'Field',
+                'ownerId': 'player-001',
+                'name': 'Test Card',
+                'power': Decimal('100'),
+                'damage': Decimal('50')
+            }
+        ],
+        'players': [
+            {
+                'id': 'player-001',
+                'name': 'Player 1',
+                'leaderId': 'leader-001'
+            },
+            {
+                'id': 'player-002',
+                'name': 'Player 2',
+                'leaderId': 'leader-002'
+            }
+        ],
+        'battleStep': 'BlockChoice',
+        'pendingBattle': {
+            'attackerId': 'card-001',
+            'attackerOwnerId': 'player-001'
+        },
+        'turnPlayerId': 'player-001',
+        'status': 'InProgress',
+        'updatedAt': '2024-01-01T00:00:00Z'
+    }
+    
+    table.put_item(Item=test_match)
+    
+    # テストケース 1: 無効な攻撃者ID
+    context = {'field': 'declareAttack', 'args': {'attackerId': 'invalid-card'}}
+    event = {'info': context, 'arguments': context['args']}
+    
+    result = lambda_handler(event, None)
+    
+    # エラーイベントが返されることを確認
+    assert 'events' in result
+    assert len(result['events']) == 1
+    assert result['events'][0]['type'] == 'InvalidAttacker'
+    assert 'message' in result['events'][0]['payload']
+    print("✓ 無効な攻撃者ID テスト通過")
+    
+    # テストケース 2: 攻撃者IDが存在しない
+    context = {'field': 'declareAttack', 'args': {}}
+    event = {'info': context, 'arguments': context['args']}
+    
+    result = lambda_handler(event, None)
+    
+    # 空のイベントが返されることを確認
+    assert 'events' in result
+    assert len(result['events']) == 0
+    print("✓ 攻撃者ID未指定 テスト通過")
+    
+    # テストケース 3: 無効なカードID (summonCard)
+    context = {'field': 'summonCard', 'args': {'cardId': 'invalid-card'}}
+    event = {'info': context, 'arguments': context['args']}
+    
+    result = lambda_handler(event, None)
+    
+    # エラーイベントが返されることを確認
+    assert 'events' in result
+    assert len(result['events']) == 1
+    assert result['events'][0]['type'] == 'CardNotFound'
+    assert 'message' in result['events'][0]['payload']
+    print("✓ 無効なカードID テスト通過")
+    
+    # テストケース 4: マッチIDが存在しない
+    context = {'field': 'getMatch', 'args': {}}
+    event = {'info': context, 'arguments': context['args']}
+    
+    result = lambda_handler(event, None)
+    
+    # エラーイベントが返されることを確認
+    assert 'events' in result
+    assert len(result['events']) == 1
+    assert result['events'][0]['type'] == 'MissingMatchId'
+    assert 'message' in result['events'][0]['payload']
+    print("✓ マッチID未指定 テスト通過")
+    
+    # テストケース 5: サポートされていないフィールド
+    context = {'field': 'unsupportedField', 'args': {'matchId': 'test-match-001'}}
+    event = {'info': context, 'arguments': context['args']}
+    
+    result = lambda_handler(event, None)
+    
+    # エラーイベントが返されることを確認
+    assert 'events' in result
+    assert len(result['events']) == 1
+    assert result['events'][0]['type'] == 'UnsupportedField'
+    assert 'message' in result['events'][0]['payload']
+    print("✓ サポート外フィールド テスト通過")
+    
+    print("\n🎉 すべてのテストが通過しました！")
+
+
+if __name__ == '__main__':
+    test_safe_error_handling()

